File: Tracker/src/test/java/com/cyrillrx/tracker/TestUtils.java
package com.cyrillrx.tracker;

import com.cyrillrx.tracker.context.TrackerContext;
import com.cyrillrx.tracker.event.TrackEvent;
import com.cyrillrx.tracker.utils.Utils;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

/**
 * @author Cyril Leroux
 *         Created on 25/04/16.
 */
public class TestUtils {

    public static final String EVENT_CATEGORY = "ScreenView";
    public static final String EVENT_NAME = "Home screen view";
    public static final String EVENT_SOURCE = "AppClassOrScreen";
    public static final String EVENT_ID = "home";
    public static final String EVENT_TYPE = "screen";

    public static final String KEY_CUSTOM_1 = "key_custom_1";
    public static final String KEY_CUSTOM_2 = "key_custom_2";

    public static TrackerContext createFakeContext() {
        return new TrackerContext()
                .setApp(new TrackerContext.App("AwesomeApp", "2.0"))
                .setUser(new TrackerContext.User()
                        .setId("john_doe")
                        .setName("John Doe")
                        .setEmail("john_doe@company.com"))
                .setDevice(new TrackerContext.Device());
    }

    static TrackEvent createFakeEvent(String category) {
        return new TrackEvent.Builder().setCategory(category).build();
    }

    static void trackEventsOneByOne(TrackerChild tracker, int count, String category) {
        for (int i = 0; i < count; i++) {
            tracker.track(createFakeEvent(category));
        }
    }

    static void trackEventsBatch(TrackerChild tracker, int count, String category) {

        final List<TrackEvent> eventBucket = new ArrayList<>();
        for (int i = 0; i < count; i++) {
            eventBucket.add(createFakeEvent(category));
        }
        tracker.track(eventBucket);
    }

    static void wait100Millis() { Utils.wait(100, TimeUnit.MILLISECONDS); }
}

File: tracker/src/test/java/com/cyrillrx/tracker/TestUtils.kt
package com.cyrillrx.tracker

import com.cyrillrx.tracker.context.TrackerContext
import com.cyrillrx.tracker.context.TrackerContext.App
import com.cyrillrx.tracker.context.TrackerContext.Device
import com.cyrillrx.tracker.event.TrackEvent
import com.cyrillrx.tracker.utils.waitFor
import kotlinx.coroutines.delay
import org.junit.Assert
import java.util.ArrayList
import java.util.concurrent.TimeUnit
import kotlin.random.Random

/**
 * @author Cyril Leroux
 * Created on 25/04/2016.
 */
object TestUtils {

    const val EVENT_CATEGORY = "ScreenView"
    const val EVENT_NAME = "Home screen view"
    const val EVENT_SOURCE = "AppClassOrScreen"

    const val KEY_1 = "key_custom_1"
    const val KEY_2 = "key_custom_2"

    const val VALUE_1 = "hello"
    const val VALUE_2 = "world"

    fun createFakeContext(): TrackerContext {
        return TrackerContext()
            .setApp(App("AwesomeApp", "2.0"))
            .setUser(
                TrackerContext.User()
                    .setId("john_doe")
                    .setName("John Doe")
                    .setEmail("john_doe@company.com")
            )
            .setDevice(Device())
    }

    fun createFakeEvent(category: String): TrackEvent =
        TrackEvent.Builder()
            .setCategory(category)
            .setName(Random(10).nextInt(10).toString())
            .build()

    suspend fun trackEventsOneByOne(tracker: TrackerChild, count: Int, category: String) {
        for (i in 0 until count) {
            tracker.track(createFakeEvent(category))
        }
        delay(10L)
    }

    suspend fun trackEventsBatch(tracker: TrackerChild, count: Int, category: String) {
        val eventBucket: MutableList<TrackEvent> = ArrayList()
        for (i in 0 until count) {
            eventBucket.add(createFakeEvent(category))
        }
        tracker.track(eventBucket)
        delay(10L)
    }

    fun wait100Millis() {
        waitFor(100, TimeUnit.MILLISECONDS)
    }

    fun assertTrackEventConsistency(event: TrackEvent) {
        Assert.assertEquals("Event category is inconsistent.", EVENT_CATEGORY, event.category)
        Assert.assertEquals("Event name is inconsistent.", EVENT_NAME, event.name)
        Assert.assertEquals("Event source is inconsistent.", EVENT_SOURCE, event.source)

        Assert.assertEquals("Custom attr 1 is inconsistent.", VALUE_1, event.getCustomAttribute(KEY_1))
        Assert.assertEquals("Custom attr 2 is inconsistent.", VALUE_2, event.getCustomAttribute(KEY_2))
    }
}